### PR TITLE
[SPARK-35244][SQL] Invoke should throw the original exception

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR updates the interpreted code path of invoke expressions, to unwrap the `InvocationTargetException`

### Why are the changes needed?

Make interpreted and codegen path consistent for invoke expressions.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new UT

### CODE REVIEW

**Comment by cloud-fan** [comment]
@viirya @maropu

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by viirya — COMMENTED** [review]
lgtm

**Comment by maropu** [comment]
late lgtm.

**Comment by cloud-fan** [comment]
I've backported this to 3.1/3.0, as inconsistency between codegen and interpreted code path is risky.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -129,7 +129,12 @@ trait InvokeLike extends Expression with NonSQLExpression {
       // return null if one of arguments is null
       null
     } else {
-      val ret = method.invoke(obj, args: _*)
+      val ret = try {
+        method.invoke(obj, args: _*)
+      } catch {
+        // Re-throw the original exception.
+        case e: java.lang.reflect.InvocationTargetException => throw e.getCause
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:136]: nit: a potential NPE here:)

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:136]: do you mean `e.getCause` can be null?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:136]: right

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:136]: I don't know how the cause can be null for `InvocationTargetException`, but let me add the null check for safety.

**Pull request closed without merging**